src/systems/filecoin_mining/storage_mining/challenge_status.go

```diff
@@ -1,25 +1,53 @@
 package storage_mining
 
 import (
+	"math"
+
 	block "github.com/filecoin-project/specs/systems/filecoin_blockchain/struct/block"
 )
 
-func (cs *ChallengeStatus_I) OnNewChallenge(currEpoch block.ChainEpoch) ChallengeStatus {
+// update pointer to most recent challenge
+func (cs *ChallengeStatus_I) OnNewChallenge(currEpoch block.ChainEpoch) {
 	cs.LastChallengeEpoch_ = currEpoch
-	return cs
 }
 
-// Call by _onSuccessfulPoSt or _onMissedSurprisePoSt
-func (cs *ChallengeStatus_I) OnChallengeResponse(currEpoch block.ChainEpoch) ChallengeStatus {
-	cs.LastChallengeEndEpoch_ = currEpoch
-	return cs
+// Update pointer to most recent successful challenge response (both ePoSt and sPoSt)
+// Call by _onSuccessfulPoSt
+func (cs *ChallengeStatus_I) OnPoStSuccess(currEpoch block.ChainEpoch) {
+	cs._lastPoStSuccessEpoch_ = currEpoch
+}
+
+// Update pointer to most recent challenge response failure
+// Call by  _onMissedSurprisePoSt
+func (cs *ChallengeStatus_I) OnPoStFailure(currEpoch block.ChainEpoch) {
+	cs._lastPoStFailureEpoch_ = currEpoch
+}
+
+func (cs *ChallengeStatus_I) LastPoStResponseEpoch() block.ChainEpoch {
+	return block.ChainEpoch(math.Max(float64(cs._lastPoStSuccessEpoch()), float64(cs._lastPoStFailureEpoch())))
 }
 
 func (cs *ChallengeStatus_I) IsChallenged() bool {
-	// true (isChallenged) when LastChallengeEpoch is later than LastChallengeEndEpoch
-	return cs.LastChallengeEpoch() > cs.LastChallengeEndEpoch()
+	// true if most recent challenge has gone unanswered
+	return cs.LastChallengeEpoch() > cs.LastPoStResponseEpoch()
+}
+
+func (cs *ChallengeStatus_I) ChallengeHasExpired(currEpoch block.ChainEpoch) bool {
+	// check if current challenge is past due
+	// TODO: pull in from consts
+	PROVING_PERIOD := block.ChainEpoch(0)
+	return cs.IsChallenged() && currEpoch > cs.LastChallengeEpoch()+PROVING_PERIOD
+}
+
+func (cs *ChallengeStatus_I) CanBeElected(currEpoch block.ChainEpoch) bool {
+	// true if most recent successful post (surprise or election) was recent enough
+	// and not currently getting challenged
+
+	// TODO: pull in from consts
```

Comment:
there is a `constants.go` in storage_mining that this can live in

Reply:
yeah I know. I just don't know how we should go about pulling things in from there (dependency wise). It may be trivial.